doc/reference/reference_lua/config/utils_schema.rst

```diff
@@ -0,0 +1,1163 @@
+
+**Since:** :doc:`3.2.0 </release/3.2.0>`
+
+The ``experimental.config.utils.schema`` module is used to validate and process
+parts of cluster configurations that have arbitrary user-defined structures:
+
+-   :ref:`app.cfg <configuration_reference_app_cfg>` for applications loaded using the :ref:`app <configuration_reference_app>` option
+-   :ref:`roles_cfg <configuration_reference_roles_cfg>` for :ref:`custom roles <application_roles>` developed as a part of a cluster application
+
+The module provides an API to get and set configuration values, filter and transform configuration data, and so on.
+
+..  important::
+
+    ``experimental.config.utils.schema`` is an experimental submodule and is subject to changes.
+
+
+.. _config_utils_schema_getting_started:
+
+Getting started with config.utils.schema
+----------------------------------------
+
+As an example, consider an :ref:`application role <configuration_application_roles>`
+that has a single configuration option - an HTTP endpoint address.
+
+..  literalinclude:: /code_snippets/snippets/config/instances.enabled/config_schema_nodes_scalar/config.yaml
+    :language: yaml
+    :start-at: roles
+    :dedent:
+
+This is how you can use the ``experimental.config.utils.schema`` module to process
+the role configuration:
+
+1.  Load the module:
+
+    ..  literalinclude:: /code_snippets/snippets/config/instances.enabled/config_schema_nodes_scalar/http_api.lua
+        :language: lua
+        :start-at: config.utils.schema
+        :end-at: config.utils.schema
+        :dedent:
+
+2.  Define a *schema* -- the root object that stores information about the role's
+    configuration -- using :ref:`schema.new() <config-utils-schema-new>`. The example
+    below shows a schema that includes a single string option:
+
+    ..  literalinclude:: /code_snippets/snippets/config/instances.enabled/config_schema_nodes_scalar/http_api.lua
+        :language: lua
+        :start-at: local http_api_schema
+        :end-before: local function validate
+        :dedent:
+
+3.  Use the :ref:`validate() <config-schema_object-validate>` function to
+    validate configuration values against the schema. In case of a role, call this
+    function inside the role's :ref:`validate() <roles_create_custom_role_validate>` function:
+
+    ..  literalinclude:: /code_snippets/snippets/config/instances.enabled/config_schema_nodes_scalar/http_api.lua
+        :language: lua
+        :start-at: local function validate
+        :end-before: local function apply
+        :dedent:
+
+4.  Obtain values of configuration options using :ref:`get() <config-schema_object-get>`.
+    In case of a role, call it inside the role's :ref:`apply() <roles_create_custom_role_apply>` function:
+
+    ..  literalinclude:: /code_snippets/snippets/config/instances.enabled/config_schema_nodes_scalar/http_api.lua
+        :language: lua
+        :start-at: local function apply
+        :end-before: local function stop
+        :dedent:
```

Comment:
The code here is the following:

```lua
local function apply(cfg)
    local http_api_cfg = http_api_schema:get(cfg)
    log.info("HTTP API endpoint: %s", http_api_cfg)
end
```

But there is no sense to call `<schema object>:get(data, path)` if the `path` is `nil`. It returns the same `cfg` value in this case.

There is some sense to use `:get()` when we have two nesting levels in the configuration, for example:

```lua
local bar = myschema:get(cfg, 'foo.bar')
```

Unlike `cfg.foo.bar` here we have the optional chaining semantic: if there is no `foo`, we get `nil`, instead of `error: attempt to index nil value`.

Even using `:get()` with just one level (with a constant path) has no sense:

```lua
local foo = myschema:get(cfg, 'foo')
-- roughtly same as
local foo = cfg.foo
```

The former checks that the `foo` field exists in the schema, but the value is anyway constant.

Reply:
Replaced the code with a link to a correct example below.